framework/src/araui/libs/upng.c

```diff
@@ -0,0 +1,1299 @@
+				out[(obp) >> 3] |= (1 << (7 - ((obp) & 0x7)));
+			++obp;
+		}
+		ibp += diff;
+	}
+}
+
+/*out must be buffer big enough to contain full image, and in must contain the full decompressed data from the IDAT chunks*/
+static void post_process_scanlines(upng_t* upng, unsigned char *out, unsigned char *in, const upng_t* info_png)
+{
+	unsigned bpp = upng_get_bpp(info_png);
+	unsigned w = info_png->width;
+	unsigned h = info_png->height;
+
+	if (bpp == 0) {
+		SET_ERROR(upng, UPNG_EMALFORMED);
+		return;
+	}
+
+	if (bpp < 8 && w * bpp != ((w * bpp + 7) / 8) * 8) {
+		unfilter(upng, in, in, w, h, bpp);
+		if (upng->error != UPNG_EOK) {
+			return;
+		}
+		remove_padding_bits(out, in, w * bpp, ((w * bpp + 7) / 8) * 8, h);
+	} else {
+		unfilter(upng, out, in, w, h, bpp);	/*we can immediatly filter into the out buffer, no other steps needed */
+	}
+}
+
+static upng_format determine_format(upng_t* upng) {
+	switch (upng->color_type) {
+	case UPNG_LUM:
+		switch (upng->color_depth) {
+		case 1:
+			return UPNG_LUMINANCE1;
+		case 2:
+			return UPNG_LUMINANCE2;
+		case 4:
+			return UPNG_LUMINANCE4;
+		case 8:
+			return UPNG_LUMINANCE8;
+		default:
+			return UPNG_BADFORMAT;
+		}
+	case UPNG_RGB:
+		switch (upng->color_depth) {
+		case 8:
+			return UPNG_RGB8;
+		case 16:
+			return UPNG_RGB16;
+		default:
+			return UPNG_BADFORMAT;
+		}
+	case UPNG_LUMA:
+		switch (upng->color_depth) {
+		case 1:
+			return UPNG_LUMINANCE_ALPHA1;
+		case 2:
+			return UPNG_LUMINANCE_ALPHA2;
+		case 4:
+			return UPNG_LUMINANCE_ALPHA4;
+		case 8:
+			return UPNG_LUMINANCE_ALPHA8;
+		default:
+			return UPNG_BADFORMAT;
+		}
+	case UPNG_RGBA:
+		switch (upng->color_depth) {
+		case 8:
+			return UPNG_RGBA8;
+		case 16:
+			return UPNG_RGBA16;
+		default:
+			return UPNG_BADFORMAT;
+		}
+	default:
+		return UPNG_BADFORMAT;
+	}
+}
+
+static void upng_free_source(upng_t* upng)
+{
+	if (upng->source.owning != 0) {
+		free((void*)upng->source.buffer);
+	}
+
+	upng->source.buffer = NULL;
+	upng->source.size = 0;
+	upng->source.owning = 0;
+}
+
+/*read the information from the header and store it in the upng_Info. return value is error*/
+upng_error upng_header(upng_t* upng)
+{
+	/* if we have an error state, bail now */
+	if (upng->error != UPNG_EOK) {
+		return upng->error;
+	}
+
+	/* if the state is not NEW (meaning we are ready to parse the header), stop now */
+	if (upng->state != UPNG_NEW) {
+		return upng->error;
+	}
+
+	/* minimum length of a valid PNG file is 29 bytes
+	 * FIXME: verify this against the specification, or
+	 * better against the actual code below */
+	if (upng->source.size < 29) {
+		SET_ERROR(upng, UPNG_ENOTPNG);
+		return upng->error;
+	}
+
+	/* check that PNG header matches expected value */
+	if (upng->source.buffer[0] != 137 || upng->source.buffer[1] != 80 || upng->source.buffer[2] != 78 || upng->source.buffer[3] != 71 || upng->source.buffer[4] != 13 || upng->source.buffer[5] != 10 || upng->source.buffer[6] != 26 || upng->source.buffer[7] != 10) {
```

Comment:
I think you'd better explain regarding buffer including information of each position of buffer.

Reply:
Same as above